File: tdd-katas/video-store-kata/video-store-python/src/customer.py
from src.movie import MovieType

class Customer:
  def __init__(self, name):
    self._name = name
    self._rentals = []

  def add_rental(self, rental): 
    self._rentals.append (rental)
	
  def get_name (self):
    return self._name
	
  def statement (self):
    total_amount 			= 0
    frequent_renter_points 	= 0
    result 					= "Rental Record for " + self.get_name () + "\n"
		
    for rental in self._rentals:
      this_amount = 0
			
      # determines the amount for each line
      movie_type = rental.get_movie ().get_price_code () 
      if movie_type == MovieType.REGULAR:
        this_amount += 2
        if rental.get_days_rented() > 2:
          this_amount += (rental.get_days_rented() - 2) * 1.5
      elif movie_type ==	MovieType.NEW_RELEASE:
        this_amount += rental.get_days_rented() * 3
      elif movie_type == MovieType.CHILDRENS:
        this_amount += 1.5
        if rental.get_days_rented() > 3:
          this_amount += (rental.get_days_rented() - 3) * 1.5
			
      frequent_renter_points += 1
			
      if rental.get_movie ().get_price_code () == MovieType.NEW_RELEASE and rental.get_days_rented() > 1:
        frequent_renter_points += 1
				
      result += "\t" + rental.get_movie ().get_title () + "\t" + "{:.1f}".format(this_amount) + "\n"
      total_amount += this_amount
		
    result += "You owed " + "{:.1f}".format(total_amount) + "\n"
    result += "You earned " + str (frequent_renter_points) + " frequent renter points\n"		
		
    return result

File: tdd-katas/video-store-kata/video-store-typescript/src/Customer.ts
export enum MovieType {
  REGULAR = "Regular movie",
  NEW_RELEASE = "New release movie",
  CHILDRENS = "Childrens movie",
}

export class Movie {
  constructor(
    public readonly title: string,
    public readonly movieType: MovieType
  ) {}
}

export class Rental {
  constructor(public readonly movie: Movie, public readonly days: number) {}
}

export class Customer {
  constructor(private name: string, private rentals: Rental[] = []) {}

  public addRental(rental: Rental): void {
    this.rentals.push(rental);
  }

  public statement(): string {
    let totalAmount = 0;
    let frequentRenterPoints = 0;
    let result = `Rental Record for ${this.name}\n`;

    for (let rental of this.rentals) {
      let thisAmount = 0;

      // determine amount for each movie
      switch (rental.movie.movieType) {
        case MovieType.REGULAR:
          thisAmount = 2;
          if (rental.days > 2) {
            thisAmount += (rental.days - 2) * 1.5;
          }
          break;
        case MovieType.NEW_RELEASE:
          thisAmount = rental.days * 3;
          break;
        case MovieType.CHILDRENS:
          thisAmount = 1.5;
          if (rental.days > 3) {
            thisAmount += (rental.days - 3) * 1.5;
          }
          break;
      }

      //add frequent renter points
      frequentRenterPoints++;
      // add bonus for a two day new release rental
      if (rental.movie.movieType === MovieType.NEW_RELEASE && rental.days > 2)
        frequentRenterPoints++;

      //print figures for this rental
      result += `\t${rental.movie.title}\t${thisAmount}\n`;
      totalAmount += thisAmount;
    }
    // add footer lines
    result += `Amount owed is ${totalAmount}\n`;
    result += `You earned ${frequentRenterPoints} frequent renter points\n`;

    return result;
  }
}
